File: ios/RNChangeBundleLib.m
#import <React/RCTBridgeModule.h>
#import "RNChangeBundleLib.h"
#import "RNChangeBundleFS.h"

static NSString * const activeBundleName = @"activeBundle";
static NSString * const nameBundleList = @"bundleList";
static NSString * const nameWaitingReactStart = @"waitReactStart";
static NSString * const nameNativeBuildVersion = @"nativeBuildVersion";
static NSString * const nameShouldDropActiveVersion = @"shouldDropActiveVersion";


@interface RNChangeBundleLib () <RCTBridgeModule>
@end

static NSURL *_defaultBundleURL = nil;

@implementation RNChangeBundleLib {

}

+ (void)initialize
{
    [super initialize];
}

+ (NSURL *)bundleURL
{
    NSMutableDictionary *dict = [RNChangeBundleFS loadStore];
    
    // Если стоит дефолтный запуск
    if ([dict[activeBundleName] isEqualToString: @""]){
        return _defaultBundleURL;
    } else {
        
        //Если прошло нативное обновление или не указано
        if ([dict[nameNativeBuildVersion] isEqualToString: @""] || [dict[nameNativeBuildVersion] isEqualToString: [RNChangeBundleLib getBuildId]]){
            // Если нативного обновления не было
            // Если стоит запуск кастомного бандла
            NSString *path = [RNChangeBundleFS getBundleFileNameForBundleId:dict[activeBundleName]];
            
            // Проверка на прошлый запуск реакта
            if ([dict[nameWaitingReactStart] boolValue]){
                // Если прошлый раз реакт не стартанул
                dict[nameWaitingReactStart] = @NO;
                dict[nameShouldDropActiveVersion] = @YES;
                [RNChangeBundleFS saveStore:dict];
                return _defaultBundleURL;
            } else {
                // Если прошлый раз реакт стартанул
                BOOL isFileExists = [RNChangeBundleFS exists:path];
                // Проверка на наличие этого кастомного файла
                if (isFileExists){
                    // Если файл существует, то запускаем проверку на успешный старт реакта
                    dict[nameWaitingReactStart] = @YES;
                    [RNChangeBundleFS saveStore:dict];
                    return [NSURL fileURLWithPath:path];
                } else {
                    // Если файла нет, то и кастомного реакта нет
                    dict[nameShouldDropActiveVersion] = @YES;
                    [RNChangeBundleFS saveStore:dict];
                    return _defaultBundleURL;
                }
            }
        } else {
            // Если было нативное обновление
            dict[nameShouldDropActiveVersion] = @YES;
            [RNChangeBundleFS saveStore:dict];
            return _defaultBundleURL;
        }
    }
}




+ (void)setDefaultBundleURL:(NSURL *)URL
{
    _defaultBundleURL = URL;
}


+ (NSError *)addBundle:(NSString *)bundleId pathForBundle:(NSString *)bundlePath pathForAssets:(NSString *)assetsPath
{
    NSError *error = nil;
    
    NSString *path = [RNChangeBundleFS getFolderForBundleId:bundleId];

    BOOL isFolderExists = [RNChangeBundleFS exists:path];
    
    if (!isFolderExists){
        error = [RNChangeBundleFS createFolder:path];

        if (error != nil){
            NSLog(@"error %@", error);
            return error;
        }
    }
    
    error = [RNChangeBundleFS moveWithOverride:bundlePath to:[RNChangeBundleFS getBundleFileNameForBundleId:bundleId]];
    
    if (error != nil){
        NSLog(@"error %@", error);
        return error;
    }
    
    error = [RNChangeBundleFS moveWithOverride:assetsPath to:[RNChangeBundleFS getAssetsFolderNameForBundleId:bundleId]];
    
    if (error != nil){
        NSLog(@"error %@", error);
        return error;
    }
    
    NSMutableDictionary *dict = [RNChangeBundleFS loadStore];
    dict[nameBundleList][bundleId] = bundleId;
    [RNChangeBundleFS saveStore:dict];
    return error;
}

- (void)addBundlePromise:(NSString *)bundleId pathForBundle:(NSString *)bundlePath pathForAssets:(NSString *)assetsPath withResolver: (RCTPromiseResolveBlock)resolve
                  withRejecter: (RCTPromiseRejectBlock)reject
{
    NSError *error = nil;
    error = [RNChangeBundleLib addBundle:bundleId pathForBundle:bundlePath pathForAssets:assetsPath];
    
    if (error == nil){
        resolve(@"Added");
    } else {
        reject(@"RNChangeBundle", error.localizedDescription, error);
    }
}

+ (NSError *)deleteBundle:(NSString *)bundleId
{
    NSError *error = nil;
    
    NSString *path = [RNChangeBundleFS getFolderForBundleId:bundleId];

    BOOL isFolderExists = [RNChangeBundleFS exists:path];
    
    if (isFolderExists){
        error = [RNChangeBundleFS remove:path];

        if (error != nil){
            NSLog(@"error %@", error);
            return error;
        }
    }
    
    
    NSMutableDictionary *dict = [RNChangeBundleFS loadStore];
    [dict[nameBundleList] removeObjectForKey:bundleId];
    [RNChangeBundleFS saveStore:dict];
    return error;
}

- (void)deleteBundlePromise:(NSString *)bundleId withResolver: (RCTPromiseResolveBlock)resolve
                  withRejecter: (RCTPromiseRejectBlock)reject
{
    NSError *error = nil;
    
    error = [RNChangeBundleLib deleteBundle:bundleId];
    
    if (error == nil){
        resolve(@"Deleted");
    } else {
        reject(@"RNChangeBundle", error.localizedDescription, error);
    }
}

+ (NSDictionary *)getBundles
{
    NSMutableDictionary *dict = [RNChangeBundleFS loadStore];
    return dict[nameBundleList];
}

- (void)getBundlesPromise:(RCTPromiseResolveBlock)resolve
                  withRejecter: (RCTPromiseRejectBlock)reject
{
    resolve([RNChangeBundleLib getBundles]);
}

+ (NSString *)getActiveBundle
{
    NSMutableDictionary *dict = [RNChangeBundleFS loadStore];
    return dict[activeBundleName];
}

- (void)getActiveBundlePromise:(RCTPromiseResolveBlock)resolve
                  withRejecter: (RCTPromiseRejectBlock)reject
{
    resolve([RNChangeBundleLib getActiveBundle]);
}

+ (void)activateBundle:(NSString *)bundleId
{
    NSMutableDictionary *dict = [RNChangeBundleFS loadStore];
    dict[activeBundleName] = bundleId;
    dict[nameNativeBuildVersion] = [RNChangeBundleLib getBuildId];
    [RNChangeBundleFS saveStore:dict];
}

- (void)activateBundlePromise:(NSString *)bundleId withResolver: (RCTPromiseResolveBlock)resolve
                  withRejecter: (RCTPromiseRejectBlock)reject
{
    [RNChangeBundleLib activateBundle:bundleId];
    
    resolve(@"Activated");
}

+ (void)notifyIfUpdateApplies
{
    NSMutableDictionary *dict = [RNChangeBundleFS loadStore];
    dict[nameWaitingReactStart] = @NO;
    [RNChangeBundleFS saveStore:dict];
    if ([dict[nameShouldDropActiveVersion] boolValue]){
        [RNChangeBundleLib deleteBundle:dict[activeBundleName]];
        [RNChangeBundleLib activateBundle:@""];
        NSMutableDictionary *dict = [RNChangeBundleFS loadStore];
        dict[nameShouldDropActiveVersion] = @NO;
        dict[nameNativeBuildVersion] = @"";
        [RNChangeBundleFS saveStore:dict];
    }
}

- (void)notifyIfUpdateAppliesPromise:(RCTPromiseResolveBlock)resolve
                  withRejecter: (RCTPromiseRejectBlock)reject
{
    [RNChangeBundleLib notifyIfUpdateApplies];
    resolve(@"Notified");
}

+ (void)reload
{
    if ([NSThread isMainThread]) {
        RCTTriggerReloadCommandListeners(@"react-native-restart: Restart");
    } else {
        dispatch_sync(dispatch_get_main_queue(), ^{
            RCTTriggerReloadCommandListeners(@"react-native-restart: Restart");
        });
    }
}

- (void)reloadPromise:(RCTPromiseResolveBlock)resolve
        withRejecter: (RCTPromiseRejectBlock)reject
{
    [RNChangeBundleLib reload];
    resolve(@"Reloaded");
}

+ (NSString *) getBuildId {
    return [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
}

- (void) getBuildIdPromise:(RCTPromiseResolveBlock)resolve
              withRejecter: (RCTPromiseRejectBlock)reject {
    resolve([RNChangeBundleLib getBuildId]);
}


RCT_REMAP_METHOD(addBundle, exportedAddBundle:(NSString *)bundleId pathForBundle:(NSString *)bundlePath  pathForAssets:(NSString *)assetsPath withResolver: (RCTPromiseResolveBlock)resolve
                 withRejecter: (RCTPromiseRejectBlock)reject)
{
    [self addBundlePromise:bundleId pathForBundle:bundlePath pathForAssets:assetsPath withResolver:resolve withRejecter:reject];
}

RCT_REMAP_METHOD(deleteBundle, exportedDeleteBundle:(NSString *)bundleId withResolver: (RCTPromiseResolveBlock)resolve withRejecter: (RCTPromiseRejectBlock)reject)
{
    [self deleteBundlePromise:bundleId withResolver:resolve withRejecter:reject];
}

RCT_REMAP_METHOD(getBundles, exportedGetBundles:(RCTPromiseResolveBlock)resolve withRejecter: (RCTPromiseRejectBlock)reject)
{
    [self getBundlesPromise:resolve withRejecter:reject];
}

RCT_REMAP_METHOD(getActiveBundle, exportedGetActiveBundle:(RCTPromiseResolveBlock)resolve withRejecter: (RCTPromiseRejectBlock)reject)
{
    [self getActiveBundlePromise:resolve withRejecter:reject];
}

RCT_REMAP_METHOD(activateBundle, exportedActivateBundle:(NSString *)bundleId withResolver: (RCTPromiseResolveBlock)resolve withRejecter: (RCTPromiseRejectBlock)reject)
{
    [self activateBundlePromise:bundleId withResolver:resolve withRejecter:reject];
}


RCT_REMAP_METHOD(notifyIfUpdateApplies, exportedNotifyIfUpdateApplies:(RCTPromiseResolveBlock)resolve withRejecter: (RCTPromiseRejectBlock)reject)
{
    [self notifyIfUpdateAppliesPromise:resolve withRejecter:reject];
}

RCT_REMAP_METHOD(reload, exportedReload:(RCTPromiseResolveBlock)resolve withRejecter: (RCTPromiseRejectBlock)reject)
{
    [self reloadPromise:resolve withRejecter:reject];
}

RCT_REMAP_METHOD(getBuildId, exportedGetBuildId:(RCTPromiseResolveBlock)resolve withRejecter: (RCTPromiseRejectBlock)reject)
{
    [self getBuildIdPromise:resolve withRejecter:reject];
}



RCT_EXPORT_MODULE()


@end

File: android/src/main/java/ru/nasvyazi/changebundle/RNChangeBundleLib.java
package ru.nasvyazi.changebundle;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.util.Log;

import com.jakewharton.processphoenix.ProcessPhoenix;

import java.util.HashSet;
import java.util.Set;

public class RNChangeBundleLib {
    static String nameBundleList = "bundleList";
    static String activeBundleName = "activeBundle";
    static String nameWaitingReactStart = "waitReactStart";
    static String nameNativeBuildVersion = "nativeBuildVersion";
    static String nameShouldDropActiveVersion = "shouldDropActiveVersion";

    public static String getBundleURL(Context context) {
        SharedPreferences prefs = RNChangeBundleFS.getPreferences(context);
        String activeBundle = prefs.getString(activeBundleName, "");
        String nativeBuildVersion = prefs.getString(nameNativeBuildVersion, "");
        boolean waitReactStart = prefs.getBoolean(nameWaitingReactStart, false);

        // Если стоит дефолтный запуск
        if (activeBundle.equals("")){

            return null;
        } else {

            //Если прошло нативное обновление или не указано
            if (nativeBuildVersion.equals("") || nativeBuildVersion.equals(RNChangeBundleLib.getBuildId(context))){
                // Если нативного обновления не было
                // Если стоит запуск кастомного бандла

                String path = RNChangeBundleFS.getBundleFileNameForBundleId(context, activeBundle);

                // Проверка на прошлый запуск реакта
                // if (waitReactStart){
                // Временное решение
                if (false){
                    // Если прошлый раз реакт не стартанул
                    RNChangeBundleFS.commitOnPreferences(context, editor -> {
                        editor.putBoolean(nameWaitingReactStart, false);
                        editor.putBoolean(nameShouldDropActiveVersion, true);
                    });

                    return null;
                } else {

                    // Если прошлый раз реакт стартанул
                    boolean isFileExists = RNChangeBundleFS.exists(path);
                    // Проверка на наличие этого кастомного файла
                    if (isFileExists){

                        boolean isFileNotChanged = RNChangeBundleFS.verifyFileInfo(context, path);

                        if (isFileNotChanged){

                            // Если файл существует и не менялся, то запускаем проверку на успешный старт реакта
                            RNChangeBundleFS.commitOnPreferences(context, editor -> {
                                editor.putBoolean(nameWaitingReactStart, true);
                            });

                            return path;
                        } else {
                            // Если файл менялся, то включим ка мы дефолт
                            RNChangeBundleFS.commitOnPreferences(context, editor -> {
                                editor.putBoolean(nameShouldDropActiveVersion, true);
                            });
                            return null;
                        }
                    } else {
                        // Если файла нет, то и кастомного реакта нет
                        RNChangeBundleFS.commitOnPreferences(context, editor -> {
                            editor.putBoolean(nameShouldDropActiveVersion, true);
                        });
                        return null;
                    }
                }
            } else {

                // Если было нативное обновление
                RNChangeBundleFS.commitOnPreferences(context, editor -> {
                    editor.putBoolean(nameShouldDropActiveVersion, true);
                });
                return null;
            }
        }
    }

    public static void addBundle(Context context, String bundleId, String bundlePath, String assetsPath) {
        String folderPath = RNChangeBundleFS.getFolderForBundleId(context, bundleId);
        boolean isFolderExists = RNChangeBundleFS.exists(folderPath);
        if (!isFolderExists){
            RNChangeBundleFS.createFolder(folderPath);
        }

        RNChangeBundleFS.moveWithOverride(bundlePath, RNChangeBundleFS.getBundleFileNameForBundleId(context, bundleId));
        RNChangeBundleFS.moveWithOverride(assetsPath, RNChangeBundleFS.getAssetsFolderNameForBundleId(context, bundleId));
        RNChangeBundleFS.extractAssets(context, RNChangeBundleFS.getAssetsFolderNameForBundleId(context, bundleId), bundleId);

        SharedPreferences prefs = RNChangeBundleFS.getPreferences(context);
        Set<String> versions = prefs.getStringSet(nameBundleList, new HashSet<String>());
        RNChangeBundleFS.commitOnPreferences(context, editor -> {
            versions.add(bundleId);
            editor.putStringSet(nameBundleList, versions);
        });
    }

    public static void deleteBundle(Context context, String bundleId) {
        String folderPath = RNChangeBundleFS.getFolderForBundleId(context, bundleId);
        boolean isFolderExists = RNChangeBundleFS.exists(folderPath);

        if (isFolderExists){
            RNChangeBundleFS.remove(folderPath);
        }

        SharedPreferences prefs = RNChangeBundleFS.getPreferences(context);
        Set<String> versions = prefs.getStringSet(nameBundleList, new HashSet<String>());
        RNChangeBundleFS.commitOnPreferences(context, editor -> {
            versions.remove(bundleId);
            editor.putStringSet(nameBundleList, versions);
        });
    }

    public static Set<String> getBundles(Context context) {
        SharedPreferences prefs = RNChangeBundleFS.getPreferences(context);
        Set<String> versions = prefs.getStringSet(nameBundleList, new HashSet<String>());
        return versions;
    }

    public static String getActiveBundle(Context context) {
        SharedPreferences prefs = RNChangeBundleFS.getPreferences(context);
        String activeBundle = prefs.getString(activeBundleName, "");
        return activeBundle;
    }

    public static void activateBundle(Context context, String bundleId) {
        SharedPreferences prefs = RNChangeBundleFS.getPreferences(context);
        RNChangeBundleFS.commitOnPreferences(context, editor -> {
            editor.putString(activeBundleName, bundleId);
            editor.putString(nameNativeBuildVersion, RNChangeBundleLib.getBuildId(context));
        });
        RNChangeBundleFS.saveFileInfo(context, RNChangeBundleFS.getBundleFileNameForBundleId(context, bundleId));
    }

    public static void notifyIfUpdateApplies(Context context){
        SharedPreferences prefs = RNChangeBundleFS.getPreferences(context);
        RNChangeBundleFS.commitOnPreferences(context, editor -> {
            editor.putBoolean(nameWaitingReactStart, false);
        });
        if (prefs.getBoolean(nameShouldDropActiveVersion, false)){
            RNChangeBundleLib.deleteBundle(context, prefs.getString(activeBundleName, ""));
            RNChangeBundleLib.activateBundle(context, "");

            RNChangeBundleFS.commitOnPreferences(context, editor -> {
                editor.putBoolean(nameShouldDropActiveVersion, false);
                editor.putString(nameNativeBuildVersion, "");
            });

        }
    }

    public static void reload(Context context){
        ProcessPhoenix.triggerRebirth(context);
    }

    public static String getBuildId(Context context)  {
        try {
            PackageInfo pInfo = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
            return pInfo.versionName;
        } catch (PackageManager.NameNotFoundException e) {
            return "";
        }
    }
}
